core/schema/schema.go

```diff
@@ -339,157 +232,217 @@ func (s *schemaView) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ref string,
+	parentMod *core.Module,
+	pipeline pipeline.Path,
+) (*UserMod, error) {
+	modMeta, err := core.ModuleMetadataFromRef(
+		ctx, s.bk, s.services, s.progSockPath, pipeline, s.platform,
+		parentMod.SourceDirectory, parentMod.SourceDirectorySubpath,
+		ref,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return s.AddModFromMetadata(ctx, modMeta, pipeline)
 }
 
-func (s *schemaView) addSchemas(schemasToAdd ...ExecutableSchema) error {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	// make a copy of the current schemas
-	separateSchemas := map[string]ExecutableSchema{}
-	for k, v := range s.separateSchemas {
-		separateSchemas[k] = v
+func (s *APIServer) GetModFromMetadata(ctx context.Context, modMeta *core.Module) (*UserMod, error) {
+	dgst, err := modMeta.BaseDigest()
+	if err != nil {
+		return nil, err
 	}
+	return s.loadModCache.Get(ctx, dgst)
+}
 
-	// add in new schemas, recursively adding dependencies
-	newSchemas := make([]ExecutableSchema, 0, len(schemasToAdd))
-	for _, newSchema := range schemasToAdd {
-		// Skip adding schema if it has already been added, higher callers
-		// are expected to handle checks that schemas with the same name are
-		// actually equivalent
-		_, ok := separateSchemas[newSchema.Name()]
-		if ok {
-			continue
-		}
-
-		newSchemas = append(newSchemas, newSchema)
-		separateSchemas[newSchema.Name()] = newSchema
+func (s *APIServer) ServeModuleToMainClient(ctx context.Context, modMeta *core.Module) error {
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return err
 	}
-	if len(newSchemas) == 0 {
-		return nil
+	if clientMetadata.ModuleCallertDigest != "" {
+		return fmt.Errorf("cannot serve module to client %s", clientMetadata.ClientID)
 	}
-	sort.Slice(newSchemas, func(i, j int) bool {
-		return newSchemas[i].Name() < newSchemas[j].Name()
-	})
 
-	// merge existing and new schemas together
-	merged, err := mergeExecutableSchemas(s.mergedSchema, newSchemas...)
+	mod, err := s.GetModFromMetadata(ctx, modMeta)
 	if err != nil {
 		return err
 	}
 
-	compiled, err := compile(merged)
+	s.clientCallMu.Lock()
+	defer s.clientCallMu.Unlock()
+	callCtx, ok := s.clientCallContext[""]
+	if !ok {
+		return fmt.Errorf("client call not found")
+	}
+	deps := append([]Mod{}, callCtx.dag.mods...)
+	deps = append(deps, mod)
+	callCtx.dag, err = newModDeps(ctx, s, deps)
 	if err != nil {
 		return err
 	}
-
-	s.separateSchemas = separateSchemas
-	s.mergedSchema = merged
-	s.compiledSchema = compiled
 	return nil
 }
 
-func (s *schemaView) resolvers() Resolvers {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	return s.mergedSchema.Resolvers()
+func (s *APIServer) RegisterFunctionCall(ctx context.Context, dgst digest.Digest, mod *UserMod, call *core.FunctionCall) error {
+	if dgst == "" {
+		return fmt.Errorf("cannot register function call with empty digest")
+	}
+
+	var dag *ModDeps
+	if mod == nil {
+		// default to just serving the core API if this is for a special case like the Go SDK codegen
```

Comment:
Is Go SDK codegen the only special case? Or do you think there may be more in the future, like stdlib? (Though I'm guessing even that we'd probably handle more like the Python SDK.)

Reply:
This comment made me realize the special case of the Go SDK doesn't even matter, so I ended up just removing this entirely. The special case existed because the Go SDK was being invoked as a pretend module, but atm that's completely unnecessary now that I think about it again.

It's possible that we might want to pretend its a module again in the future, but I think doing that should be a bigger thing where it has its own `Mod` implementation similar to `core`. That would be nice now but also functionally completely useless so not going to do it atm. Can easily be done in future if needed.

I don't think there should be any more special casing in general. Like you said the stdlib should also be able to have dependencies similar to the python sdk.